### DISCUSSION TITLE
Cache Pools in Hybrid Cache

### DISCUSSION THREAD

**Discussion by pbhandar2** [facebook/CacheLib#64] [Q&A] [answered]
Hello, 

I would like to implement a few features to use Cachelib for my research. 

- As far as I understand, cache pools are only sticky in DRAM while NVM cache is shared among cache pools. I would like to implement cache pools to incorporate the NVM cache as well. 
- Furthermore, I would like there to be custom configuration for each cache pool that could comprise of DRAM/NVM cache. For instance, configuration parameter such as the probabilistic admission policy is global for all cache pools. Setting it to 0.5 means half of requests will be admitted to NVM from all cache pools. I would like to set admission rate for each cache pool if it contains a DRAM/NVM cache. 

I wanted to ask if there is any way to know about current development status of Hybrid Cache, or talk to anyone for advice, guidance on how to approach Hybrid Cache development. 

I wanted to know if there are any big obstacles, advice or things to look out when trying to implement this feature. 

Thank you!

**Comment by therealgymmy** [comment, marked as answer]
Can you explain what you're trying to achieve with cache pools on NvmCache? Also what sort of workloads?

The reason we didn't implement pools is because they can easily lead to excessive random writes, which hurt device endurance. Note this is specifically for BlockCache (large items) which benefits greatly from sequential writes.

Typically the need for pools is to isolate workload. For this, we ask the user to control this logically using the AdmissionPolicy API for NvmCache. (Refer to this for API: https://github.com/facebook/CacheLib/blob/main/cachelib/allocator/NvmAdmissionPolicy.h)

**Reply by pbhandar2** [reply]
I am looking to replay block workloads on Cachelib. I am storing data as fixed size pages (4KB). Is there any better way in Cachelib to handle fixed size allocations? These are storage workloads from VMs running database workloads. 

I am looking to split a hybrid cache between tenants sharing the cache. Each tenant is allocated a cache pool which is currently sticky only in memory. This makes analysis and prediction of cache performance on a per tenant basis difficult. 

The admission policy would only handle admitting items into the cache but I wanted cache isolation to control more of what is being evicted. Also, admission policy can only be allocated for the entire cache and not per cache pool, am I correct?

**Reply by therealgymmy** [reply]
@pbhandar2: hey sorry for the late response. You're right that there is currently no way for CacheLib to split flash-cache into different tenant pools.

If you want to measure the effect of having different tenant pools, I suggest you start with a few separate CacheLib instances first. If you see positive effects, we can discuss building native support within hybrid cache.

What are the metrics you would be looking at? (One thin I would watch out is the impact on write endurance with multiple pools on flash as writes will no longer be sequential)
